### PR TITLE
compute: add token-controlled fuse in feedback edges

### PR DESCRIPTION
**Pull request by petrosagg**
This PR implements an alternative solution to recursive dataflow cancellation where tokened passthrough operators are rendered across all feedback edges of the dataflow.

The tokens are held by the exported sinks and indexes in the same way source and imported index tokens are.

When the token gets dropped iteration will immediatley stop, akin to having just executed a `break` instruction, and even divergent dataflows quickly shut down.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/timely-util/src/operator.rs
Comment (teskje) [inline src/timely-util/src/operator.rs:489]:
```suggestion
        let mut builder = OperatorBuilderRc::new("Fused".to_owned(), self.scope());
```

**Review by teskje — APPROVED** [review]
LGTM, thanks!

I think this deserves a test. What do you think about salvaging the one from https://github.com/MaterializeInc/materialize/pull/18442? It is meant to test active dataflow cancellation, but it really only test cancellation of divergent dataflows.

**Comment by teskje** [comment]
@aalexandrov @ggevay FYI. This will fix https://github.com/MaterializeInc/materialize/issues/16800, but you might want to leave that ticket open for tracking the hard limit.

File: src/timely-util/src/operator.rs
Comment (petrosagg) [reply src/timely-util/src/operator.rs]: Fixed

**Comment by petrosagg** [comment]
@teskje I also cherry-picked the commit adding the test from https://github.com/MaterializeInc/materialize/pull/18442

**Comment by teskje** [comment]
Thanks! We should probably change the name of this file and the comments inside it, since it's not active dataflow cancellation that we are testing, but shutdown of divergent WMR dataflows. I can push another commit with that later if you don't get to it first!

File: src/timely-util/src/operator.rs
Comment (antiguru) [inline src/timely-util/src/operator.rs:150]: I'm not in favor of the name `fused`. I think the function should represent a verb, not a noun, and fused mostly means ["some things mixed together,"](https://www.merriam-webster.com/dictionary/fuse) which is not the meaning we want to have here. I liked the old `with_token` (?) better!

Comment (antiguru) [reply src/timely-util/src/operator.rs:150]: The documentation can be more specific to what the implementation actually provides. At the moment, it will gate all data once the `token` is dropped, but still passes through progress traffic. For some dataflows, this might be surprising if they assume that dropping the token also advances the frontier to the empty frontier.

**Review by antiguru — APPROVED** [review]
I think this approach is valuable! Thanks for the PR.

I added some comments around naming and describing concepts, which I think are important to resolve before merging. The concept as a whole seems good, though.

This PR can introduce a consistency when logging errors to Sentry. At the moment, we only ever process whole batches of data, which might not be the case with this PR. For reductions, this might mean that they'll see negative accumulations and log it to Sentry. Ideally, we prevent these logs, but I'm fine if it's not part of the same PR (but please make sure it'll be part of the same release.)

**Comment by teskje** [comment]
> This PR can introduce a consistency when logging errors to Sentry. At the moment, we only ever process whole batches of data, which might not be the case with this PR. For reductions, this might mean that they'll see negative accumulations and log it to Sentry. Ideally, we prevent these logs, but I'm fine if it's not part of the same PR (but please make sure it'll be part of the same release.)

This is something I have on my plate because we also need it for making persist_source exit early, and generally for everything we do for making dataflow shutdown faster. We can just leave this PR open until we have that? Should be ready somewhen next week.

File: src/timely-util/src/operator.rs
Comment (teskje) [reply src/timely-util/src/operator.rs:150]: I like having "fuse" in the name because it evokes the notion of a thing that stops stuff from flowing through once it is triggered. (It would be nicer to have a water analogy rather than one from electricity, but I can't think of one.) "token" doesn't evoke the same associations for me, it's pretty generic. But it does have a specific meaning in the context of MZ dataflows, so I think it would be fine too.

A couple more entries to the bike shedding:
* `with_fuse`
* `with_circuit_breaker`

**Comment by petrosagg** [comment]
@antiguru @teskje I pushed a commit that changes the name to `with_token`, adds a bit more documentation,  and also makes it so the frontier immediately advances to the empty antichain when the token is dropped.

**Comment by philip-stoev** [comment]
This is from the "feature flags" Nightly job:

```
thread 'timely:work-0' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/dataflow/operators/generic/builder_raw.rs:148:9
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed::<usize, usize>
             at ./rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:212:5
   4: <timely::dataflow::operators::generic::builder_raw::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>>>::new_output_connection::<alloc::vec::Vec<(mz_repr::row::Row, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>, i64)>>
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/dataflow/operators/generic/builder_raw.rs:148:9
   5: <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>>>::new_output_connection::<alloc::vec::Vec<(mz_repr::row::Row, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>, i64)>>
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/dataflow/operators/generic/builder_rc.rs:110:29
   6: <timely::dataflow::stream::StreamCore<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>, alloc::vec::Vec<(mz_repr::row::Row, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>, i64)>> as mz_timely_util::operator::StreamExt<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>, (mz_repr::row::Row, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>, i64)>>::with_token
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/timely-util/src/operator.rs:409:36
   7: <differential_dataflow::collection::Collection<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>, mz_repr::row::Row, i64> as mz_timely_util::operator::CollectionExt<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>, mz_repr::row::Row, i64>>::with_token
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/timely-util/src/operator.rs:521:9
   8: <mz_compute::render::context::Context<timely::dataflow::scopes::child::Child<timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp>, timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>>, mz_repr::row::Row>>::render_recursive_plan
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/render/mod.rs:658:22
   9: mz_compute::render::build_compute_dataflow::<timely_communication::allocator::generic::Generic>::{closure#3}::{closure#2}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/render/mod.rs:307:38
  10: <timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp> as timely::dataflow::scopes::Scope>::scoped::<timely::order::product::Product<mz_repr::timestamp::Timestamp, differential_dataflow::dynamic::pointstamp::PointStamp<usize>>, (), mz_compute::render::build_compute_dataflow<timely_communication::allocator::generic::Generic>::{closure#3}::{closure#2}>
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/dataflow/scopes/child.rs:129:13
  11: <timely::dataflow::scopes::child::Child<timely::worker::Worker<timely_communication::allocator::generic::Generic>, mz_repr::timestamp::Timestamp> as timely::dataflow::scopes::Scope>::iterative::<differential_dataflow::dynamic::pointstamp::PointStamp<usize>, (), mz_compute::render::build_compute_dataflow<timely_communication::allocator::generic::Generic>::{closure#3}::{closure#2}>
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/dataflow/scopes/mod.rs:126:9
  12: mz_compute::render::build_compute_dataflow::<timely_communication::allocator::generic::Generic>::{closure#3}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/render/mod.rs:273:13
  13: <timely::worker::Worker<timely_communication::allocator::generic::Generic>>::dataflow_core::<mz_repr::timestamp::Timestamp, (), mz_compute::render::build_compute_dataflow<timely_communication::allocator::generic::Generic>::{closure#3}, alloc::boxed::Box<()>>
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/worker.rs:639:13
  14: mz_compute::render::build_compute_dataflow::<timely_communication::allocator::generic::Generic>
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/render/mod.rs:193:5
  15: <mz_compute::compute_state::ActiveComputeState<timely_communication::allocator::generic::Generic>>::handle_create_dataflows
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/compute_state.rs:226:13
  16: <mz_compute::compute_state::ActiveComputeState<timely_communication::allocator::generic::Generic>>::handle_compute_command
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/compute_state.rs:139:43
  17: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::handle_command
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/server.rs:414:9
  18: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::run_client
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/server.rs:369:17
  19: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::setup_channel_and_run_client
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/server.rs:321:9
  20: <mz_compute::server::Worker<timely_communication::allocator::generic::Generic>>::run
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/server.rs:181:21
  21: <mz_compute::server::Config as mz_cluster::types::AsRunnableWorker<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>::build_and_run::<timely_communication::allocator::generic::Generic>
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/compute/src/server.rs:162:9
  22: <mz_cluster::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse, timely::scheduling::activate::SyncActivator>, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>, mz_compute::server::Config, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>::build_timely::{closure#0}::{closure#1}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-04708f9e0b018c0f3-1/materialize/tests/src/cluster/src/server.rs:179:13
  23: timely::execute::execute_from::<timely_communication::allocator::generic::GenericBuilder, (), <mz_cluster::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse, timely::scheduling::activate::SyncActivator>, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>, mz_compute::server::Config, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>::build_timely::{closure#0}::{closure#1}>::{closure#0}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/execute.rs:388:22
  24: timely_communication::initialize::initialize_from::<timely_communication::allocator::generic::GenericBuilder, (), timely::execute::execute_from<timely_communication::allocator::generic::GenericBuilder, (), <mz_cluster::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse, timely::scheduling::activate::SyncActivator>, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>, mz_compute::server::Config, mz_compute_client::protocol::command::ComputeCommand, mz_compute_client::protocol::response::ComputeResponse>>::build_timely::{closure#0}::{closure#1}>::{closure#0}>::{closure#0}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/communication/src/initialize.rs:316:33
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

The query to cause the crash is:

```
CREATE VIEW wmr_01 AS WITH MUTUALLY RECURSIVE     foo (a int, b int) AS (SELECT 1, 2 UNION SELECT a, 7 FROM bar),     bar (a int) as (SELECT a FROM foo) SELECT * FROM bar;
SELECT * FROM wmr_01 
```

After you fix it, can you make sure this query is put in an SLT file -- this should have been detected in the normal CI rather than in a random test in Nightly.

**Comment by petrosagg** [comment]
I'm sorry, this was a stupid bug in my code that broke recursive rendering so all recursive queries will panic. That's why testdrive 0 failed too. I'm pushing a fix now

File: src/timely-util/src/operator.rs
Comment (tokenrove) [reply src/timely-util/src/operator.rs]: The noun form of fuse is distinct from the verb form here, so I prefer `with_fuse` over `fused`.  Circuit breakers normally can be reset, unlike fuses, so I'd prefer to avoid that here.

**Review by philip-stoev — APPROVED** [review]
I pulled again and ran a stress test with 50 concurrent divergent WMR queries. All WMR dataflows both inside cursors and inside INSERT ... SELECT are being cancelled correctly. At the end of the test computed CPU usage drops to zero, so it is all good.

**Comment by teskje** [comment]
This PR now needs to be adapted to the changes made in https://github.com/MaterializeInc/materialize/pull/18772 (specifically https://github.com/MaterializeInc/materialize/pull/18772/commits/2a9097f29e09ef5916e3777d9fda4c7eb4127605). I can take care of that if you don't mind @petrosagg.

There is some uncertainty about aggressively advancing the fuse operator's output to the empty frontier: https://github.com/MaterializeInc/materialize/pull/18760#discussion_r1168502708. I think if we cannot come to an agreement here quickly, we should merge this PR with the initial version of the fuse operator that doesn't have special handling of its output frontier, to unblock divergent WMR cancellation.

**Comment by teskje** [comment]
As the sqllogictests found, the last commit that made the `with_token` operator advance to the empty frontier broke WMR queries. I suspect something was wrong with the way that operator tried to advance its frontier in spite of having an empty path summary on its output. I decided to revert this change, which also sidesteps the "is this even safe discussion" and brings us to a PR we can merge now.

**Comment by teskje** [comment]
I think this PR is ready for the final review. All comments have been addressed AFAICT. @vmarcos and @tokenrove can you take a look please?

File: src/timely-util/src/operator.rs
Comment (vmarcos) [inline src/timely-util/src/operator.rs:151]: I think this comment is no longer accurate? The operator does not shut down per se, but rather consumes all its input without producing output.

File: src/timely-util/src/operator.rs
Comment (vmarcos) [inline src/timely-util/src/operator.rs:219]: Same remark as above concerning the comment.

File: src/timely-util/src/operator.rs
Comment (vmarcos) [inline src/timely-util/src/operator.rs:151]: I am wondering why we need this method to be in `StreamExt`? All of our uses are covered by the definition in `CollectionExt` AFAICT. No strong opinion here, just a remark regarding maintaining more code around than we might need.

File: src/timely-util/src/operator.rs
Comment (vmarcos) [inline src/timely-util/src/operator.rs:414]: I suppose this could be expressed directly as a `unary` operator, thus avoiding the Timely boilerplate?

**Review by vmarcos — APPROVED** [review]
The changes look good to me; I added a few smaller comments regarding code organization / comments, but nothing that would stand in the way of merging.

File: src/timely-util/src/operator.rs
Comment (teskje) [reply src/timely-util/src/operator.rs]: Fixed!

File: src/timely-util/src/operator.rs
Comment (teskje) [reply src/timely-util/src/operator.rs]: Fixed!

File: src/timely-util/src/operator.rs
Comment (teskje) [reply src/timely-util/src/operator.rs]: Good catch!

**Review by tokenrove — APPROVED** [review]

**Comment by teskje** [comment]
TFTRs! And thanks @petrosagg for the implementation!

**Pull request merged**
